internal/query/rolebindings.go

```diff
@@ -0,0 +1,624 @@
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		rb, err := e.GetRoleBinding(ctx, rbRes)
+		if err != nil {
+			if errors.Is(err, ErrRoleBindingNotFound) {
+				// print and record a warning message when there's a grant points
+				// to a role-binding that not longer exists.
+				//
+				// this should not happen in normal circumstances, but it's possible
+				// if some role-binding relationships are deleted directly through
+				// spiceDB
+				err := fmt.Errorf("%w: dangling grant relationship: %s", err, rel.String())
+
+				span.RecordError(err)
+				e.logger.Warnf(err.Error())
+			}
+
+			errs = append(errs, err)
+
+			continue
+		}
+
+		if optionalRole != nil && rb.Role.ID.String() != optionalRole.ID.String() {
+			continue
+		}
+
+		if len(rb.Subjects) == 0 {
+			continue
+		}
+
+		bindings = append(bindings, rb)
+	}
+
+	for _, err := range errs {
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return bindings, nil
+}
+
+func (e *engine) UpdateRoleBinding(ctx context.Context, rb types.Resource, subjects []types.RoleBindingSubject) (types.RoleBinding, error) {
+	ctx, span := e.tracer.Start(
+		ctx, "engine.UpdateRoleBindings",
+		trace.WithAttributes(
+			attribute.Stringer("rolebinding_id", rb.ID),
+		),
+	)
+	defer span.End()
+
+	rolebinding, err := e.GetRoleBinding(ctx, rb)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.RoleBinding{}, err
+	}
+
+	// 1. find the subjects to add or remove
+	current := make([]string, len(rolebinding.Subjects))
+	incoming := make([]string, len(subjects))
+
+	for i, subj := range rolebinding.Subjects {
+		current[i] = subj.SubjectResource.ID.String()
+	}
+
+	for i, subj := range subjects {
+		incoming[i] = subj.SubjectResource.ID.String()
+	}
+
+	add, remove := diff(current, incoming)
+
+	// return if there are no changes
+	if (len(add) + len(remove)) == 0 {
+		return rolebinding, nil
+	}
+
+	// 2. create relationship updates
+	updates := make([]*pb.RelationshipUpdate, len(add)+len(remove))
+	i := 0
+
+	for _, id := range add {
+		update, err := e.rolebindingRelationshipUpdateForSubject(id, rb.ID.String(), pb.RelationshipUpdate_OPERATION_TOUCH)
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			return types.RoleBinding{}, err
+		}
+
+		updates[i] = update
+		i++
+	}
+
+	for _, id := range remove {
+		update, err := e.rolebindingRelationshipUpdateForSubject(id, rb.ID.String(), pb.RelationshipUpdate_OPERATION_DELETE)
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			return types.RoleBinding{}, err
+		}
+
+		updates[i] = update
+		i++
+	}
+
+	_, err = e.client.WriteRelationships(ctx, &pb.WriteRelationshipsRequest{Updates: updates})
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.RoleBinding{}, err
+	}
+
+	rolebinding.Subjects = subjects
+
```

Comment:
As above, call `upsertZedToken` with the token returned from `WriteRelationships`.

Reply:
I added a [wrapper function `applyUpdates`](https://github.com/infratographer/permissions-api/pull/249/files#diff-8a06560f98df912ac583f2a3b001ba10320e4adaef92cf805eb703c95fa9de4bR1273-R1289) that runs the `WriteRelationships` and also runs `upsertZedToken` on each of the resources